rs/rosetta-api/icrc1/ledger/tests/tests.rs

```diff
@@ -1632,26 +1633,69 @@ mod incompatible_token_type_upgrade {
         );
         assert_eq!(approval_result, Ok(BlockIndex::from(1u64)));
         balance -= FEE;
+        let initial_total_supply = icrc1_total_supply(&env, ledger_id);
+        assert_eq!(initial_total_supply.to_u64(), Some(u64::MAX - FEE));
+        let initial_blocks = icrc3_get_blocks(
+            &env,
+            ledger_id,
+            vec![GetBlocksRequest {
+                start: Nat::from(0u64),
+                length: Nat::from(u64::MAX),
+            }],
+        );
+        assert_eq!(initial_blocks.log_length, Nat::from(2u64));
+        assert_eq!(initial_blocks.blocks.len(), 2);
+        assert_eq!(initial_blocks.archived_blocks.len(), 0);
+        let initial_metadata = metadata(&env, ledger_id);
+        assert!(!initial_metadata.is_empty());
 
         // Try to upgrade the ledger from using a u64 wasm to a u256 wasm
         let upgrade_args = Encode!(&LedgerArgument::Upgrade(None)).unwrap();
         env.upgrade_canister(ledger_id, ledger_mainnet_u256_wasm(), upgrade_args)
             .expect("Unable to upgrade the ledger canister");
 
-        // The balance and allowance should not change
-        assert_eq!(balance, balance_of(&env, ledger_id, account(1)));
-        let allowance = get_allowance(&env, ledger_id, account(1), account(2));
-        assert_eq!(allowance.allowance, Nat::from(u64::MAX));
+        // The balance, allowance, total supply, and blocks should not change
+        let verify_state = || {
+            assert_eq!(balance, balance_of(&env, ledger_id, account(1)));
+            let allowance = get_allowance(&env, ledger_id, account(1), account(2));
+            assert_eq!(allowance.allowance, Nat::from(u64::MAX));
+            assert_eq!(
+                initial_blocks,
+                icrc3_get_blocks(
+                    &env,
+                    ledger_id,
+                    vec![GetBlocksRequest {
+                        start: Nat::from(0u64),
+                        length: Nat::from(u64::MAX),
+                    }]
+                )
+            );
+            assert_eq!(initial_metadata, metadata(&env, ledger_id));
+        };
+        verify_state();
+        // The total supply is calculated based on the token type, so this actually changes (even though it should not)
```

Comment:
This comment I don't get yet. Could you pinpoint where this bug occurs?

Reply:
The value of `ledger.balances.tokens_pool` is set to `TokensType::Tokens::max_value()` when the `balances` are initialized in `LedgerBalances::default()` [here](https://sourcegraph.com/github.com/dfinity/ic@211fa36d1de895d0295ecb27aa174756e881237b/-/blob/rs/rosetta-api/ledger_core/src/balances.rs?L116-121), which is called when the ledger is initialized in `from_init_args` [here](https://sourcegraph.com/github.com/dfinity/ic@211fa36d1de895d0295ecb27aa174756e881237b/-/blob/rs/rosetta-api/icrc1/ledger/src/lib.rs?L422). Each time new tokens are minted, the `tokens_pool` value is decreased - when tokens are burned, the `tokens_pool` value is increased. When the `total_supply` method is called, the value that is returned is calculated by subtracting the value of `tokens_pool` from `TokensType::Tokens::max_value()` as shown [here](https://sourcegraph.com/github.com/dfinity/ic@211fa36d1de895d0295ecb27aa174756e881237b/-/blob/rs/rosetta-api/ledger_core/src/balances.rs?L231-238).

In this particular case, we install a ledger with a `u64` wasm, so `tokens_pool` is set to `u64::MAX`, and a call to `total_supply` would return `u64::MAX - u64::MAX` which is `0` (before any mints). If the ledger is upgraded to a `u256` wasm, `tokens_pool` would still be `u64::MAX`, but `total_supply` would calculate `u256::MAX - u64::MAX` which is much larger than `0`.

I tried to add another test where we install a ledger with a `u256` wasm and upgrade it to `u64` wasm, expecting that a call to `total_supply` would trap due to `u64::MAX.checked_sub(u256::MAX)`, but the upgrade fails (traps) already in the `post_upgrade` due to a failed deserialization of the ledger state.

Reply:
Thanks a lot @mbjorkqvist for the detailed explanation!

> Each time new tokens are minted, the tokens_pool value is decreased - when tokens are burned, the tokens_pool value is increased.

It's an interesting design choice. It's not clear to me why it was done in this way. I would have naively expected that the total supply is tracked as a "simple" counter, that would start at 0 and is increased/decresed depending on the mint/burn operations, which I would expect to work as long as the total supply `<= TokensType::Tokens::max_value()`. I find it quite unexpected that changing the token type changes the total supply (I would expect an error or the same number as before, but not a different number). I guess that's all the more reasons to prevent this type of upgrades.